src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor{TPixel}.cs

```diff
@@ -0,0 +1,168 @@
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Binarization
+{
+    /// <summary>
+    /// Performs Bradley Adaptive Threshold filter against an image.
+    /// </summary>
+    internal class AdaptiveThresholdProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        private readonly AdaptiveThresholdProcessor definition;
+        private readonly PixelOperations<TPixel> pixelOpInstance;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveThresholdProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="configuration">The configuration which allows altering default behaviour or extending the library.</param>
+        /// <param name="definition">The <see cref="AdaptiveThresholdProcessor"/> defining the processor parameters.</param>
+        /// <param name="source">The source <see cref="Image{TPixel}"/> for the current processor instance.</param>
+        /// <param name="sourceRectangle">The source area to process for the current processor instance.</param>
+        public AdaptiveThresholdProcessor(Configuration configuration, AdaptiveThresholdProcessor definition, Image<TPixel> source, Rectangle sourceRectangle)
+            : base(configuration, source, sourceRectangle)
+        {
+            this.pixelOpInstance = PixelOperations<TPixel>.Instance;
+            this.definition = definition;
+        }
+
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source)
+        {
+            var intersect = Rectangle.Intersect(this.SourceRectangle, source.Bounds());
+
+            Configuration configuration = this.Configuration;
+            TPixel upper = this.definition.Upper.ToPixel<TPixel>();
+            TPixel lower = this.definition.Lower.ToPixel<TPixel>();
+            float thresholdLimit = this.definition.ThresholdLimit;
+
+            // Used ushort because the values should never exceed max ushort value.
+            ushort startY = (ushort)intersect.Y;
+            ushort endY = (ushort)intersect.Bottom;
+            ushort startX = (ushort)intersect.X;
+            ushort endX = (ushort)intersect.Right;
+
+            ushort width = (ushort)intersect.Width;
+            ushort height = (ushort)intersect.Height;
+
+            // ClusterSize defines the size of cluster to used to check for average. Tweaked to support up to 4k wide pixels and not more. 4096 / 16 is 256 thus the '-1'
+            byte clusterSize = (byte)Math.Truncate((width / 16f) - 1);
+
+            // Using pooled 2d buffer for integer image table and temp memory to hold Rgb24 converted pixel data.
+            using (Buffer2D<ulong> intImage = this.Configuration.MemoryAllocator.Allocate2D<ulong>(width, height))
+            using (IMemoryOwner<Rgb24> tmpBuffer = this.Configuration.MemoryAllocator.Allocate<Rgb24>(width * height))
```

Comment:
We should use `Buffer2D` here also.

Reply:
the buffer does not seem to be needed. I have removed it.

Reply:
Legend